src/main/java/org/terasology/logic/health/HealthClientSystem.java

```diff
@@ -1,34 +1,71 @@
-/*
- * Copyright 2016 MovingBlocks
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
 package org.terasology.logic.health;
 
+import com.google.common.annotations.VisibleForTesting;
+import org.joml.Vector3f;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
 import org.terasology.entitySystem.systems.BaseComponentSystem;
 import org.terasology.entitySystem.systems.RegisterMode;
 import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.logic.health.event.OnDamagedEvent;
+import org.terasology.logic.location.LocationComponent;
+import org.terasology.logic.players.PlayerCharacterComponent;
+import org.terasology.math.Direction;
 import org.terasology.registry.In;
 import org.terasology.rendering.nui.NUIManager;
+import org.terasology.rendering.nui.layers.hud.DirectionalDamageOverlay;
 
 @RegisterSystem(RegisterMode.CLIENT)
 public class HealthClientSystem extends BaseComponentSystem {
 
+    private static final float DAMAGE_OVERLAY_DELAY_SECONDS = 0.5f;
+
     @In
     private NUIManager nuiManager;
 
+    private DirectionalDamageOverlay directionalDamageOverlay;
+
     @Override
     public void initialise() {
         nuiManager.getHUD().addHUDElement("healthHud");
+        directionalDamageOverlay = (DirectionalDamageOverlay) nuiManager.getHUD().addHUDElement(
+                "directionalDamageOverlay");
+    }
+
+    @ReceiveEvent(components = PlayerCharacterComponent.class)
+    public void onDamaged(OnDamagedEvent event, EntityRef entity,
+                          LocationComponent locationComponent,
+                          HealthComponent healthComponent) {
+        // Show the relevant direction element
+        EntityRef instigator = event.getInstigator();
+        if (instigator != null && instigator.hasComponent(LocationComponent.class)) {
```

Comment:
We also need to look at the cases in which `instigator` is null, e.g. for falling damage or damage due to Hunger and Thirst.
However, it's fine to do this as a follow-up PR and keep the scope of this PR to "add red flash on receiving damage from other entities"